### PR TITLE
Bug Fix - Start Docker Containers

### PR DESCRIPTION
**Pull request by zacharyt20**
$POCS/bin/pocs tries to start docker containers when they are already running. Added an if statement that exits the code if docker containers are already running.  

## Description
Compares the docker ps command output with the pocs-shell container ID to see if the pocs-shell container is running. If it finds that the container is running it exits the script before attempting to start the docker containers and outputs an error with information on how to manually kill docker containers in case a single docker container failed.

## Related Issue
[926](https://github.com/panoptes/POCS/issues/926)

## How Has This Been Tested?
Killed all docker containers, then ran ```bin/pocs up -d``` from the ```$POCS``` directory to make sure containers still start, and then ran ```bin/pocs up -d``` again which behaved as expected and in the way described in the description section.
## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: bin/pocs [hunk 1]
```diff
@@ -34,6 +34,17 @@ if [ "${START}" = 'help' ] || [ "${START}" = '-h' ] || [ "${START}" = '--help' ]
 	exit 1
 fi
 
+#Patch could be improved by using for loops to test for all docker containers
+#in case the user manually kills pocs-shell. Also the if statement should be true
+#if the string doesn't have characters
+
+#exit script before trying to start containers if containers are running
+docker_ps=$(docker ps -q -f name=pocs-shell)
+if [ "$docker_ps" = "87d6fde2a139" ]; then
```
Comment (wtgee) [inline bin/pocs]: This is hard-coded to a specific container name but is the right approach.  Basically if it returns anything other than nothing than it is running something (how's that for a vague statement :grin:). 

In `bin/pocs-cmd` we just check for an empty string via:

```bash
if [ ! "$(docker ps -q -f name=${DOCKER_NAME})" ]; then
```

That should be good enough here. If we wanted to force certain versions we could use a specific id otherwise best to be generic.

**Comment by zacharyt20** [comment]
> Note that this assumes that POCS is the only container-based system running on the computer, which might be a bad assumption but I think is ok.

Using an if statement with multiple parameters avoids using a while loop and doesn't make this assumption.  

Also, I think it needs to check if each POCS container is running. If there is an issue in the future with docker or POCS that causes only a portion of the containers to stop(fail), the method of checking only one container would likely cause ```bin/pocs``` to start containers that are already running.

**Comment by wtgee** [comment]
> Using an if statement with multiple parameters avoids using a while loop and doesn't make this assumption.
> 
> Also, I think it needs to check if each POCS container is running. If there is an issue in the future with docker or POCS that causes only a portion of the containers to stop(fail), the method of checking only one container would likely cause `bin/pocs` to start containers that are already running.

In this situation you would either want to start the individual container with `bin/pocs up --no-deps pocs-shell` or stop all containers and restart them, so I don't think that's an issue.

Also, just a note, I just checked this and running `bin/pocs up` does not restart existing containers, although it does attached the current terminal to the log output messages (unless run with `-d` of course).  This can make it awkward as then you would have two terminals running the same thing. But there shouldn't be any conflict in terms of trying to start resources again.

**Comment by zacharyt20** [comment]
I think the average user would still attempt ```bin/pocs up -d``` to start the containers, even if they realize only certain containers are running. I could keep the same if statement and explain the `bin/pocs up --no-deps pocs-shell` syntax to the user in the error message so that way they don't have to do research on their own / ask on the forum.

**Comment by wtgee** [comment]
> I could keep the same if statement and explain the bin/pocs up --no-deps pocs-shell syntax to the user in the error message so that way they don't have to do research on their own / ask on the forum.

This is already explained in the help for `bin/pocs`. So if a user runs `bin/pocs` or `bin/pocs -h` or `bin/pocs --help` it will show the explanation.

To be clear, the only issue is that if people run `bin/pocs up` in two different terminals then it is attached in two terminals, which is awkward but doesn't prevent any functionality. So we could either always force `-d`, or only force `-d` when *any* service is already running.

This latter solution looks like the following and is pretty clean.

```bash
#!/bin/bash    
 
# Get user params
PARAMS="$@"    
    
cd "$PANDIR" 
CMD="docker-compose \    
    --project-directory ${PANDIR} \    
    -f panoptes-utils/docker/docker-compose.yaml \    
    -f PAWS/docker/docker-compose.yaml \    
    -f POCS/docker/docker-compose.yaml \    
    -p panoptes"    
    
# If user only asked to start, check if already running and if so use "-d" option.
if [[ "$PARAMS" == "up" ]]; then    
    if [[ ! -z $(eval "${CMD} ps -q") ]]; then    
        echo "Some containers already running, using -d to only start non-running containers."    
        PARAMS="up -d"    
    fi    
fi    
    
# Run the docker-compose command with user params.
eval "${CMD} ${PARAMS}"
```

**Comment by zacharyt20** [comment]
I was under the impression that `bin/pocs up` and `bin/pocs up -d` did the same thing with only difference being that the `-d` command runs in the background while `pocs up` runs in the foreground and shows in-depth debug messages. How/why would they be doing different things? From my understanding of `bin/pocs up`, it seems that adding the `-d` to `bin/pocs up` only serves to hide the ugly errors in the debug messages that are spat out when `bin/pocs up` is run with containers already started.

In the google docs documentation it says to use `bin/pocs up -d`. The only reason why someone would use `bin/pocs up` is to see the additional debug messages, which this solution hides and doesn't offer a way to see those messages or point the user towards the instructions in `bin/pocs`.

It's also again making the assumption that there are no other docker containers other than the ones made by ```pocs```.  This could be problematic for future developers trying to test new hardware who are using containers with the hopes of eventually integrating with the docker version of POCS.

**Comment by wtgee** [comment]
> I was under the impression that `bin/pocs up` and `bin/pocs up -d` did the same thing with only difference being that the `-d` command runs in the background while `pocs up` runs in the foreground and shows in-depth debug messages. How/why would they be doing different things? From my understanding of `bin/pocs up`, it seems that adding the `-d` to `bin/pocs up` only serves to hide the ugly errors in the debug messages that are spat out when `bin/pocs up` is run with containers already started.

That's correct, they are doing the same thing.

> 
> In the google docs documentation it says to use `bin/pocs up -d`. The only reason why someone would use `bin/pocs up` is to see the additional debug messages, which this solution hides and doesn't offer a way to see those messages or point the user towards the instructions in `bin/pocs`.

I use `bin/pocs up` because I prefer to keep the logs running so I can be sure that I have started the containers. Therefore I've never actually had the problem of trying to run `bin/pocs up` twice (with or without the `-d`) because I can clearly see that it's still running.

Also, this solution does not in fact do as you mention. It only forces the `-d` option if there are containers already running, in which case you do not want to use just `up` by itself, which is the actual issue this PR is trying to solve.  If no containers are running you can use just `bin/pocs up` to view log message and attach to the terminal just fine.

As it also says in the `bin/pocs --help` message, you can view the logs at any time with `bin/pocs logs <service-name>`, whether attached or not.

> It's also again making the assumption that there are no other docker containers other than the ones made by `pocs`. This could be problematic for future developers trying to test new hardware who are using containers with the hopes of eventually integrating with the docker version of POCS.

This is using the `docker-compose` version of `ps`, not the `docker` version, so it only reports on the containers that are defined in the various `docker-compose.yaml` files that are specified. (You can get it to report all running containers with `-a` but that is not the default). Other stand-alone containers won't be reported and won't affect the operation. I tested this all before I posted the solution.

**Comment by zacharyt20** [comment]
I have been assuming that the debug messages that I saw in terminal with `bin/pocs up` were also happening in the `bin/pocs up -d` log file (Note: I don't have the log file mentioned in the help section of `bin/pocs` on PAN015). I had tested `docker-compose ps` on our unit but since it only has containers within `$PANDIR` I didn't realize it filtered docker containers so cleanly!

I would also like to clarify that the log file does not show the errors that appear in `bin/pocs up` since they are not on my unit or their location is different. 

My only small concern is that someone completely new to bash/programming will try starting the containers for the first or second time and try to see the `bin/pocs up` debug messages in terminal, and won't know how to get that working. In the newest commit I've added another echo line pointing them toward the help section with docker commands.

File: bin/pocs [hunk 2]
```diff
@@ -25,6 +25,12 @@ usage() {
 
     # Run the software tests (no hardware)
     $POCS/scripts/pocs-docker.sh up
+    
+    #Start specific docker containers in the back ground with
+    $POCS/scripts/pocs-docker.sh up --no-deps -d <container name>
```
Comment (wtgee) [inline bin/pocs]: Should maintain consistency in what we call
```suggestion
    $POCS/bin/pocs up --no-deps -d <container name> 
```

File: bin/pocs [hunk 3]
```diff
@@ -34,11 +40,24 @@ if [ "${START}" = 'help' ] || [ "${START}" = '-h' ] || [ "${START}" = '--help' ]
...
+    -f panoptes-utils/docker/docker-compose.yaml \
+    -f PAWS/docker/docker-compose.yaml \
+    -f POCS/docker/docker-compose.yaml \
+    -p panoptes"
+    
+# If user only asked to start, check if already running and if so use "-d" option.
+if [[ "$PARAMS" == "up" ]]; then
+    if [[ ! -z $(eval "${CMD} ps -q") ]]; then
+        echo "Some containers already running, using -d to only start non-running containers."
+	echo "For more info on managing docker containers manually, run bin/pocs."
```
Comment (wtgee) [inline bin/pocs]:
```suggestion
	 echo "For more info on managing docker containers manually, run bin/pocs --help".
```

**Comment by zacharyt20** [comment]
Thanks for helping me with my first PR!

**Comment by zacharyt20** [comment]
I applied the changes to `bin/pocs` on our unit and it is not working. It looks like the output of `docker-compose` does not change if the docker containers are running or stopped. This causes the second if statement to always be true, and the user can never see `bin/pocs` in the foreground and gets the "Some containers are already running" message when there are no containers running.

File: bin/pocs [hunk 4]
```diff
@@ -34,11 +40,24 @@ if [ "${START}" = 'help' ] || [ "${START}" = '-h' ] || [ "${START}" = '--help' ]
...
+CMD="docker-compose \
+    --project-directory ${PANDIR} \
+    -f panoptes-utils/docker/docker-compose.yaml \
+    -f PAWS/docker/docker-compose.yaml \
+    -f POCS/docker/docker-compose.yaml \
+    -p panoptes"
+    
+# If user only asked to start, check if already running and if so use "-d" option.
+if [[ "$PARAMS" == "up" ]]; then
+    if [[ ! -z $(eval "${CMD} top") ]]; then
```
Comment (wtgee) [inline bin/pocs]: Why the change to `top` from `ps -q`? I think either will work, just curious.

Comment (zacharyt20) [reply bin/pocs]: I replaced `ps -q` with `top` to filter for running processes. 

On our system, when we tried `bin/pocs up` with **no** containers running, the bash script added the `-d` option, which is not the intended behavior.  We echoed the parameter of the if statement (`$(eval "${CMD} ps -q")`), and saw that this parameter is the always the same, and does not change based on the status of the docker containers. My best guess as to why this happens is that `docker ps` lists running containers, while `docker-compose ps` lists any information it can get from the `docker-compose.yaml` files, which it can find when docker containers aren't running.

Using `top` lists only the running processes.

Comment (wtgee) [reply bin/pocs]: Hm, that doesn't happen for me, `ps -q` will only give output. I guess maybe if you have some stopped machines? I tend to do `docker system prune` a lot as it cleans up a lot of hard drive space, but that would remove stopped and unused containers.

**Comment by wtgee** [comment]
Will be closed by #951

**Comment by wtgee** [comment]
As per comment above, I'm closing this now as part of my cleanup effort.
